site/app/controllers/pdf/PDFController.php

```diff
@@ -14,6 +14,20 @@ public function __construct(Core $core) {
         parent::__construct($core);
     }
 
+    public function getRealPath(string $file_path, string $id): string {
```

Comment:
Should this function throw an exception if the file doesn't exist, or if you don't have access to the file?  At a minimum, it should probably check for `..` in strings.  This isn't an active security issue but it might be good to future-proof things a bit so we don't accidentally add a security bug in the future.